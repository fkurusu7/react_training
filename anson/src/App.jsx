import { useEffect, useState } from "react";
import JPPosts from "./components/JPPosts";
import JPUser from "./components/JPUser";
import UserContext from "./utils/context/userContext";
import { useFetchUser } from "./hooks/useFetchUser";

/* import { useState } from "react";
import UserItem from "./components/UserItem";
 */
// export default function App({ usersData }) {
export default function App() {
  /* const [users, setUsers] = useState(usersData);

  return (
    <div>
      {users.map((user) => (
        <UserItem user={user} setUsers={setUsers} key={user.id} />
      ))}
    </div>
  ); */
  const { userData, error, isLoading } = useFetchUser(2);
  const [user, setUser] = useState({});

  useEffect(() => {
    if (!isLoading && !error && userData) setUser(userData);
  }, [userData, error, isLoading]);

  return (
    <UserContext.Provider value={{ ...user, setUser }}>
      <div>
        <h1>JSON PlaceHolder Users</h1>
        <div>
          {isLoading ? "Loading..." : <JPUser />}
          {!error && <p>{error}</p>}
        </div>
        <JPPosts />
      </div>
    </UserContext.Provider>
  );
}
