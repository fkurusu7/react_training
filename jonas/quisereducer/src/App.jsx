import { useEffect, useReducer } from 'react';

import Error from './components/Error';
import FinishScreen from './components/FinishScreen';
import Footer from './components/Footer';
import Header from './components/Header';
import Loader from './components/Loader';
import Main from './components/Main';
import NextButton from './components/NextButton';
import Progress from './components/Progress';
import Question from './components/Question';
import StartScreen from './components/StartScreen';
import Timer from './components/Timer';
import { QUESTIONS_TYPE_ACTIONS, STATUS_DATA } from './utils/constants';

const QUESTIONS_URI = 'http://localhost:3000/questions';

const initialState = {
  questions: [],
  status: STATUS_DATA.Loading,
  index: 0,
  answer: null,
  points: 0,
  highscore: 0,
  secondsRemaining: null,
};

const SECONDS_PER_QUESTION = 30;
function quizReducer(state, action) {
  switch (action.type) {
    case QUESTIONS_TYPE_ACTIONS.DataReceived:
      return { ...state, questions: action.payload, status: STATUS_DATA.Ready };

    case QUESTIONS_TYPE_ACTIONS.DataFailed:
      return { ...state, status: STATUS_DATA.Error };

    case QUESTIONS_TYPE_ACTIONS.Start:
      return {
        ...state,
        status: STATUS_DATA.Active,
        secondsRemaining: state.questions.length * SECONDS_PER_QUESTION,
      };

    case QUESTIONS_TYPE_ACTIONS.NewAnswer: {
      const currentQuestion = state.questions.at(state.index);
      const stateAnswer = action.payload;

      return {
        ...state,
        answer: stateAnswer,
        points:
          stateAnswer === currentQuestion.correctOption
            ? state.points + currentQuestion.points
            : state.points,
      };
    }
    case QUESTIONS_TYPE_ACTIONS.NextQuestion:
      return { ...state, index: state.index + 1, answer: null };

    case QUESTIONS_TYPE_ACTIONS.Finish: {
      const highscore = Math.max(state.highscore, state.points);
      return { ...state, status: STATUS_DATA.Finished, highscore };
    }

    case QUESTIONS_TYPE_ACTIONS.Restart:
      return {
        ...initialState,
        status: STATUS_DATA.Ready,
        questions: state.questions,
        highscore: state.highscore,
      };

    case QUESTIONS_TYPE_ACTIONS.Tick:
      return {
        ...state,
        secondsRemaining: state.secondsRemaining - 1,
        status:
          state.secondsRemaining === 0 ? STATUS_DATA.Finished : state.status,
      };

    default:
      throw new Error('Action Unknown');
  }
}

function App() {
  // state
  const [state, dispatch] = useReducer(quizReducer, initialState);
  const {
    status,
    questions,
    index,
    answer,
    points,
    highscore,
    secondsRemaining,
  } = state;
  const numQuestions = questions.length;
  const maxPoints = questions.reduce(
    (acc, currQuestion) => acc + currQuestion.points,
    0
  );

  useEffect(() => {
    const abortController = new AbortController();

    async function fetchQuestions() {
      try {
        const response = await fetch(QUESTIONS_URI, {
          signal: abortController.signal,
        });
        const data = await response.json();

        dispatch({ type: QUESTIONS_TYPE_ACTIONS.DataReceived, payload: data });
      } catch (error) {
        console.log(error);
        dispatch({ type: QUESTIONS_TYPE_ACTIONS.DataFailed });
      }
    }

    fetchQuestions();

    return () => {
      abortController.abort();
    };
  }, []);

  console.log({ status, questions, answer, points, highscore });

  return (
    <div className='app'>
      <Header />
      <Main>
        {status === STATUS_DATA.Loading && <Loader />}
        {status === STATUS_DATA.Error && <Error />}
        {status === STATUS_DATA.Ready && (
          <StartScreen numQuestions={numQuestions} dispatch={dispatch} />
        )}
        {status === STATUS_DATA.Active && (
          <>
            <Progress
              index={index}
              numQuestions={numQuestions}
              points={points}
              maxPoints={maxPoints}
              answer={answer}
              highscore={highscore}
            />
            <Question
              question={questions[index]}
              dispatch={dispatch}
              answer={answer}
            />
            <Footer>
              <Timer dispatch={dispatch} secondsRemaining={secondsRemaining} />
              <NextButton
                dispatch={dispatch}
                answer={answer}
                numQuestions={numQuestions}
                index={index}
              />
            </Footer>
          </>
        )}
        {status === STATUS_DATA.Finished && (
          <FinishScreen
            points={points}
            maxPossiblePoints={maxPoints}
            highscore={highscore}
            dispatch={dispatch}
          />
        )}
      </Main>
    </div>
  );
}

export default App;
