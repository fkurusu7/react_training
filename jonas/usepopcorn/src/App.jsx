/* eslint-disable no-unused-vars */
import { useEffect, useRef, useState } from 'react';
import StarRating from './components/StarRating';
import { useLocalStorageState } from './hooks/useLocalStorageState';
import { useMovies } from './hooks/useMovies';

const tempMovieData = [
  {
    imdbID: 'tt1375666',
    Title: 'Inception',
    Year: '2010',
    Poster:
      'https://m.media-amazon.com/images/M/MV5BMjAxMzY3NjcxNF5BMl5BanBnXkFtZTcwNTI5OTM0Mw@@._V1_SX300.jpg',
  },
  {
    imdbID: 'tt0133093',
    Title: 'The Matrix',
    Year: '1999',
    Poster:
      'https://m.media-amazon.com/images/M/MV5BNzQzOTk3OTAtNDQ0Zi00ZTVkLWI0MTEtMDllZjNkYzNjNTc4L2ltYWdlXkEyXkFqcGdeQXVyNjU0OTQ0OTY@._V1_SX300.jpg',
  },
  {
    imdbID: 'tt6751668',
    Title: 'Parasite',
    Year: '2019',
    Poster:
      'https://m.media-amazon.com/images/M/MV5BYWZjMjk3ZTItODQ2ZC00NTY5LWE0ZDYtZTI3MjcwN2Q5NTVkXkEyXkFqcGdeQXVyODk4OTc3MTY@._V1_SX300.jpg',
  },
];

const tempWatchedData = [
  {
    imdbID: 'tt1375666',
    Title: 'Inception',
    Year: '2010',
    Poster:
      'https://m.media-amazon.com/images/M/MV5BMjAxMzY3NjcxNF5BMl5BanBnXkFtZTcwNTI5OTM0Mw@@._V1_SX300.jpg',
    runtime: 148,
    imdbRating: 8.8,
    userRating: 10,
  },
  {
    imdbID: 'tt0088763',
    Title: 'Back to the Future',
    Year: '1985',
    Poster:
      'https://m.media-amazon.com/images/M/MV5BZmU0M2Y1OGUtZjIxNi00ZjBkLTg1MjgtOWIyNThiZWIwYjRiXkEyXkFqcGdeQXVyMTQxNzMzNDI@._V1_SX300.jpg',
    runtime: 116,
    imdbRating: 8.5,
    userRating: 9,
  },
];

const average = (arr) =>
  arr.reduce((acc, cur, i, arr) => acc + cur / arr.length, 0);

export const OMDB_API_URI = 'http://www.omdbapi.com/?apikey=bfb73349';
const WATCHED_LS_KEY = 'watched';

/***************************/
/***************************/
/***************************/
/* COMPONENTS */
/***************************/

function Logo() {
  return (
    <div className='logo'>
      <span role='img'>üçø</span>
      <h1>usePopcorn</h1>
    </div>
  );
}

function Search({ query, setQuery }) {
  const searchInputRef = useRef(null);

  useEffect(() => {
    const addKeydownEvent = (ev) => {
      if (document.activeElement === searchInputRef.current) return;

      if (ev.code === 'Enter') {
        searchInputRef.current.focus();
        setQuery('');
      }
    };
    document.addEventListener('keydown', addKeydownEvent);

    return () => document.removeEventListener('keydown', addKeydownEvent);
  }, [setQuery]);

  return (
    <input
      className='search'
      type='text'
      placeholder='Search movies...'
      value={query}
      onChange={(e) => setQuery(e.target.value)}
      ref={searchInputRef}
    />
  );
}

function NumResults({ movies }) {
  return (
    <p className='num-results'>
      Found <strong>{movies?.length}</strong> results
    </p>
  );
}

function Navbar({ children }) {
  return (
    <nav className='nav-bar'>
      <Logo />
      {children}
    </nav>
  );
}

function MovieDetails({ selectedId, onCloseMovie, onAddWatched, watched }) {
  const [movie, setMovie] = useState({});
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState(undefined);
  const [userRating, setUserRating] = useState(0);

  const isWatched = watched.map((movie) => movie.imdbID).includes(selectedId);
  const watchedUserRating = watched.find(
    (movie) => movie.imdbID === selectedId
  )?.userRating;

  const {
    Title: title,
    Year: year,
    Poster: poster,
    Runtime: runtime,
    imdbRating,
    Plot: plot,
    Released: released,
    Actors: actors,
    Director: director,
    Genre: genre,
  } = movie;

  useEffect(() => {
    const abortController = new AbortController();

    async function getMovieDetails() {
      try {
        setIsLoading(true);
        const response = await fetch(`${OMDB_API_URI}&i=${selectedId}`, {
          signal: abortController.signal,
        });
        if (!response.ok) {
          throw new Error('Error fetching Movie Details');
        }

        const data = await response.json();
        setMovie(data);
      } catch (error) {
        console.error(error, error.message);
        if (error.name !== 'AbortError') {
          setError(error.message);
        }
      } finally {
        setIsLoading(false);
        setError(undefined);
      }
    }

    getMovieDetails();
    return () => {
      abortController.abort();
    };
  }, [selectedId]);

  useEffect(() => {
    if (!title) return;
    document.title = `Movie | ${title}`;

    return () => (document.title = 'Use Popcorn');
  }, [title]);

  useEffect(
    function () {
      const handleKeydown = function (ev) {
        if (ev.code === 'Escape') {
          onCloseMovie();
        }
      };

      document.addEventListener('keydown', handleKeydown);

      return function () {
        document.removeEventListener('keydown', handleKeydown);
      };
    },
    [onCloseMovie]
  );

  function handleAdd() {
    const newWatchedMovie = {
      imdbID: selectedId,
      title,
      year,
      poster,
      imdbRating: Number(imdbRating),
      runtime: Number(runtime.split(' ').at(0)),
      userRating,
    };

    onAddWatched(newWatchedMovie);
    onCloseMovie();
  }

  return (
    <div className='details'>
      {isLoading ? (
        <Loader />
      ) : (
        <>
          <header>
            <button
              className='btn-back'
              onClick={() => {
                onCloseMovie();
              }}
            >
              &larr;
            </button>

            <img src={poster} alt={`Poster of ${title} movie`} />
            <div className='details-overview'>
              <h2>{title}</h2>
              <p>
                {released} &bull; {runtime}
              </p>
              <p>{genre}</p>
              <p>
                <span>‚≠êÔ∏è</span>
                {imdbRating} IMDB Rating
              </p>
            </div>
          </header>
          <section>
            <div className='rating'>
              {isWatched ? (
                <p>
                  You rated this movie <span>{watchedUserRating}</span>
                </p>
              ) : (
                <>
                  <StarRating
                    size={24}
                    maxRating={10}
                    onSetRating={setUserRating}
                  />
                  {userRating > 0 && (
                    <button className='btn-add' onClick={handleAdd}>
                      + Add to List
                    </button>
                  )}
                </>
              )}
            </div>
            <p>
              <em>{plot}</em>
            </p>
            <p>Starring {actors}</p>
            <p>Directed by {director}</p>
          </section>
        </>
      )}
    </div>
  );
}

function Movie({ movie, onSelectMovie }) {
  return (
    <li key={movie.imdbID} onClick={() => onSelectMovie(movie.imdbID)}>
      <img src={movie.Poster} alt={`${movie.Title} poster`} />
      <h3>{movie.Title}</h3>
      <div>
        <p>
          <span>üóì</span>
          <span>{movie.Year}</span>
        </p>
      </div>
    </li>
  );
}

function MoviesList({ movies, onSelectMovie }) {
  return (
    <ul className='list list-movies'>
      {movies?.map((movie) => (
        <Movie key={movie.imdbID} movie={movie} onSelectMovie={onSelectMovie} />
      ))}
    </ul>
  );
}

function Box({ children }) {
  const [isOpen, setIsOpen] = useState(true);
  return (
    <div className='box'>
      <button className='btn-toggle' onClick={() => setIsOpen((open) => !open)}>
        {isOpen ? '‚Äì' : '+'}
      </button>
      {isOpen && children}
    </div>
  );
}

function Summary({ watched }) {
  const avgImdbRating = average(watched.map((movie) => movie.imdbRating));
  const avgUserRating = average(watched.map((movie) => movie.userRating));
  const avgRuntime = average(watched.map((movie) => movie.runtime));

  return (
    <div className='summary'>
      <h2>Movies you watched</h2>
      <div>
        <p>
          <span>#Ô∏è‚É£</span>
          <span>{watched.length} movies</span>
        </p>
        <p>
          <span>‚≠êÔ∏è</span>
          <span>{avgImdbRating.toFixed(1)}</span>
        </p>
        <p>
          <span>üåü</span>
          <span>{avgUserRating.toFixed(1)}</span>
        </p>
        <p>
          <span>‚è≥</span>
          <span>{avgRuntime} min</span>
        </p>
      </div>
    </div>
  );
}

function WatcheMovie({ movie, handleRemoveMovie }) {
  return (
    <li key={movie.imdbID}>
      <img src={movie.Poster} alt={`${movie.Title} poster`} />
      <h3>{movie.Title}</h3>
      <div>
        <p>
          <span>‚≠êÔ∏è</span>
          <span>{movie.imdbRating}</span>
        </p>
        <p>
          <span>üåü</span>
          <span>{movie.userRating}</span>
        </p>
        <p>
          <span>‚è≥</span>
          <span>{movie.runtime} min</span>
        </p>
      </div>
      <button
        type='button'
        className='btn-delete'
        onClick={() => handleRemoveMovie(movie.imdbID)}
      >
        X
      </button>
    </li>
  );
}

function WatchedMoviesList({ watched, handleRemoveMovie }) {
  return (
    <ul className='list'>
      {watched.map((movie) => (
        <WatcheMovie
          key={movie.imdbID}
          movie={movie}
          handleRemoveMovie={handleRemoveMovie}
        />
      ))}
    </ul>
  );
}

function Main({ children }) {
  return <main className='main'>{children}</main>;
}

function Loader() {
  return <p className='loader'>Loading...</p>;
}

function ErrorMessage({ error }) {
  return <p className='error'>{error}</p>;
}

function App() {
  const [query, setQuery] = useState('');

  const [selectedId, setSelectedId] = useState(null);
  // const [watched, setWatched] = useState([]);
  const { value: watched, setValue: setWatched } = useLocalStorageState(
    [],
    WATCHED_LS_KEY
  );

  const { movies, isLoading, error } = useMovies(
    query,
    handleCloseSelectedMovie
  );

  function handleSelectMovie(id) {
    setSelectedId((selectedId) => (id === selectedId ? null : id));
  }

  function handleCloseSelectedMovie() {
    setSelectedId(null);
  }

  function handleAddWatched(movie) {
    setWatched((watched) => [...watched, movie]);
  }

  function handleRemoveWatchedMovie(movieId) {
    setWatched((currWatchedMovies) =>
      currWatchedMovies.filter((watched) => watched.imdbID !== movieId)
    );
  }

  /*   useEffect(() => {
    const abortController = new AbortController();

    const fetchMovies = async () => {
      try {
        setIsLoading(true);
        setError(undefined);
        const response = await fetch(`${OMDB_API_URI}&s=${query}`, {
          signal: abortController.signal,
        });
        if (!response.ok) throw new Error('Something went wrong');
        const jsonRes = await response.json();

        if (jsonRes.Response === 'False') {
          throw new Error('Movie not found');
        }

        setMovies(jsonRes.Search);
      } catch (error) {
        if (error.name !== 'AbortError') {
          setError(error.message);
        }
      } finally {
        setIsLoading(false);
        setError(undefined);
      }
    };

    if (query.length <= 2) {
      setMovies([]);
      setError(undefined);
      return;
    }

    handleCloseSelectedMovie();
    fetchMovies();

    return () => {
      abortController.abort();
    };
  }, [query]);
 */

  return (
    <>
      <Navbar>
        <Search query={query} setQuery={setQuery} />
        <NumResults movies={movies} />
      </Navbar>
      <Main>
        <Box>
          {isLoading && <Loader />}
          {!isLoading && !error && (
            <MoviesList movies={movies} onSelectMovie={handleSelectMovie} />
          )}
          {error && <ErrorMessage error={error} />}
        </Box>
        <Box>
          {selectedId ? (
            <MovieDetails
              selectedId={selectedId}
              onCloseMovie={handleCloseSelectedMovie}
              onAddWatched={handleAddWatched}
              watched={watched}
            />
          ) : (
            <>
              <Summary watched={watched} />
              <WatchedMoviesList
                watched={watched}
                handleRemoveMovie={handleRemoveWatchedMovie}
              />
            </>
          )}
        </Box>
      </Main>
    </>
  );
}

export default App;

/* 
Component Composition:
  - Build better layouts
  - Resolve Prop Drilling
  - Create reusable components

  Implicit:
    <Box>
      <MoviesList movies={movies} />
    </Box>

  function Box({ children }) {
    const [isOpen, setIsOpen] = useState(true);
    return (
      <div className='box'>
        <button className='btn-toggle' onClick={() => setIsOpen((open) => !open)}>
          {isOpen ? '‚Äì' : '+'}
        </button>
        {isOpen && children}
      </div>
  );
}

  Explicit:
    <Box element={<MoviesList movies={movies} />} />
    <Box element={
        <>
          <Summary watched={watched} />
          <WatchedMoviesList watched={watched} />
        </>
      }
    />

    function Box({ element }) {
      const [isOpen, setIsOpen] = useState(true);
      return (
        <div className='box'>
          <button className='btn-toggle' onClick={() => setIsOpen((open) => !open)}>
            {isOpen ? '‚Äì' : '+'}
          </button>
          {isOpen && element}
        </div>
    );
}
*/
