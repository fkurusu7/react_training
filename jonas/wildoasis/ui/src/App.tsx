import styled from 'styled-components';
import GlobalStyles from './styles/GlobalStyles';
import Button from './ui/Button';
import Heading from './ui/Heading';
import Input from './ui/Input';
import Row from './ui/Row';

const StyledApp = styled.div`
  background-color: red;
  padding: 20px;
`;

function App() {
  return (
    <>
      <GlobalStyles />
      <StyledApp>
        <Row>
          <Row type='horizontal'>
            <Heading as='h1'>The Wild Oasis</Heading>
            <div>
              <Heading as='h3'>Checccck</Heading>

              <Input
                type='number'
                placeholder='# Guests'
                id='testnum'
                name='testnum'
              />
              <Button size='small' variation='danger'>
                dsds
              </Button>
            </div>
          </Row>

          <Row type='vertical'>
            <Heading as='h3'>Form</Heading>
            <form>
              <Input
                type='text'
                placeholder='Testing Row'
                id='testtext1'
                name='testtext1'
              />
              <Input
                type='text'
                placeholder='Testing Row'
                id='testtext2'
                name='testtext2'
              />
              <Button size='large' variation='primary'>
                dsds
              </Button>
            </form>
          </Row>
        </Row>
      </StyledApp>
    </>
  );
}

export default App;
