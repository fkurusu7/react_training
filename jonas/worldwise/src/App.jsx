import { BrowserRouter, Navigate, Route, Routes } from 'react-router-dom';
import City from './components/City';
import CityList from './components/CityList';
import CountryList from './components/CountryList';
import Form from './components/Form';
import { AuthProvider } from './contexts/AuthContext';
import { CitiesProvider } from './contexts/CitiesContext';

import { lazy, Suspense } from 'react';
// import AppLayout from './pages/AppLayout';
// import Homepage from './pages/Homepage';
// import Login from './pages/Login';
// import PageNotFound from './pages/PageNotFound';
// import Pricing from './pages/Pricing';
// import Product from './pages/Product';
import SpinnerFullPage from './components/SpinnerFullPage';
import ProtectedRoute from './pages/ProtectedRoute';

const AppLayout = lazy(() => import('./pages/AppLayout'));
const Homepage = lazy(() => import('./pages/Homepage'));
const Login = lazy(() => import('./pages/Login'));
const PageNotFound = lazy(() => import('./pages/PageNotFound'));
const Pricing = lazy(() => import('./pages/Pricing'));
const Product = lazy(() => import('./pages/Product'));

function App() {
  return (
    <AuthProvider>
      <CitiesProvider>
        <BrowserRouter>
          <Suspense fallback={<SpinnerFullPage />}>
            <Routes>
              <Route path='/' element={<Homepage />} />
              <Route path='products' element={<Product />} />
              <Route path='pricing' element={<Pricing />} />
              <Route path='login' element={<Login />} />

              <Route
                path='app'
                element={
                  <ProtectedRoute>
                    <AppLayout />
                  </ProtectedRoute>
                }
              >
                <Route index element={<Navigate to='cities' replace />} />
                <Route path='cities' element={<CityList />} />
                <Route path='cities/:id' element={<City />} />
                <Route path='countries' element={<CountryList />} />
                <Route path='form' element={<Form />} />
              </Route>
              <Route path='*' element={<PageNotFound />} />
            </Routes>
          </Suspense>
        </BrowserRouter>
      </CitiesProvider>
    </AuthProvider>
  );
}

export default App;
