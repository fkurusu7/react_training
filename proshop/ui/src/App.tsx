import { BrowserRouter, Route, Routes } from 'react-router-dom';
import AppLayout from './pages/AppLayout';
import HomePage from './pages/HomePage';
import ProductPage from './pages/ProductPage';

function App() {
  return (
    <BrowserRouter>
      <Routes>
        <Route path='/' element={<AppLayout />}>
          <Route index element={<HomePage />} />
          <Route path='/product/:id' element={<ProductPage />} />
        </Route>
      </Routes>
    </BrowserRouter>
  );
}

export default App;
