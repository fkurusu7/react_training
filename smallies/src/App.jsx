import ProjectCard from "./components/ProjectCard";
import projects from "./data";

const App = () => {
  return (
    <section className="container">
      <h1 className="heading heading-1">Small Projects</h1>

      <section className="projects">
        {projects.map((project) => (
          <ProjectCard
            name={project.name}
            description={project.description}
            component={project.component}
          />
        ))}
      </section>
    </section>
  );
};

export default App;
