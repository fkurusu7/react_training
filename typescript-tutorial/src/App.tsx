function App() {
  let name: string = 'Mina';
  name = 'Typescript';
  // name = 93; // Type 'number' is not assignable to type 'string'.ts(2322)

  const messageInfered = 'Turururoriri message Infered (String)';

  const numbers: number[] = [1, 2, 3, 4, 5];
  const boolArr: Array<boolean> = [true, false, false, true, true];
  const strArr: Array<string> = ['uno', 'dos', 'tres', 'cuatro', 'cinco'];

  let unionValue: string | number = '1';
  unionValue = 1;

  function printId(id: string | number): string {
    return `Your id is ${id}`;
  }

  function processValue(input: unknown): string {
    return `Unkown forces you using type guards (typeof, instanceof, etc.) before you can access its properties.- Your input: ${input} ${typeof input}`;
  }

  const carOne: { car: string; brand: number } = {
    car: 'Evil Cafè',
    brand: 2302,
  };
  // carOne.color = '#123'; // Property 'color' does not exist on type '{ car: string; brand: number; }'.ts(2339)

  const tomato = { name: 'Tomato', price: 2 };
  const potato = { name: 'Potato', price: 1 };
  const carrot = { name: 'Carrot' };

  const vegetables: readonly {
    readonly name: string;
    readonly price?: number;
  }[] = [tomato, potato, carrot];

  // vegetables.pop(); // Property 'pop' does not exist on type 'readonly { readonly name: string; readonly price?: number | undefined; }[]'.ts(2339)

  function createEmployee({ id }: { id: number }): {
    id: number;
    isActive: boolean;
  } {
    return { id, isActive: id % 2 === 0 };
  }

  type UserInfo = { name: string; age: number; address: string };
  type Address = { city: string; country: string };
  // Intersection of User and Address
  type UserWithAddress = UserInfo & Address;

  function getUserInfo(userInfo: UserWithAddress) {
    console.log(`User Info: 
    Name: ${userInfo.name}, 
    Age: ${userInfo.age}, 
    Address: ${userInfo.address},
    City: ${userInfo.city},
    Country: ${userInfo.country}`);
  }

  const user: UserWithAddress = {
    name: 'Alice',
    age: 30,
    address: '123 Main St',
    country: 'Uganda',
    city: 'Kampala',
  };
  getUserInfo(user);

  interface Pet {
    readonly id: number;
    breed: string;
    age: number;
    color: string;
  }
  interface Can extends Pet {
    name: string;
    legs: number;
    walk(steps: number, minutes: number): number;
  }

  const walk: Can['walk'] = (steps, minutes) => steps / minutes;
  console.log(walk(1000, 20));

  return (
    <div className='container'>
      <section>
        <h1>Hola {name.toUpperCase()}</h1>
        <h4>Nono: {messageInfered.toLowerCase()}</h4>

        {numbers.map((number) => (
          <span key={number}>{number} </span>
        ))}
        <br />
        {boolArr.map((bool, idx) => (
          <span key={idx + 1}>X {bool.toString()}</span>
        ))}
        <br />
        {strArr.map((str) => (
          <span key={str}> {str} </span>
        ))}
      </section>

      <section>
        <h2>UNIONS</h2>
        <p>let unionValue: string | number = '1';</p>
        <p>
          {' '}
          unionValue = 1; --- typeof {unionValue}={typeof unionValue}
        </p>

        <p>Union type in function params: </p>
        <code>function printId(id: string | number): string</code>
        <br />
        <code> return `Your id is $id`;</code>
        <p>{printId('19238')}</p>
        <hr />
        <p>Unkown Type</p>
        <p>{processValue(true)}</p>
        <p>{processValue('tu soñador')}</p>
        <p>{processValue(34)}</p>
      </section>

      <section>
        <h1>OBJECTS</h1>
        <p>
          To define the shape of an object explicitly, we can use inline type
          annotations.
        </p>

        <span>
          const carOne: &#123; car: string; brand: number &#125; = &#123; car:
          'Evil Cafè', brand: 2302, &#125;
        </span>
        <p>
          {carOne.car} - {carOne.brand}
        </p>
        <span>
          const carOne: &#123; car: string; brand: number &#125; = &#123; car:
          'Evil Cafè', brand: '2302', &#125;
        </span>
        <p style={{ color: '#f86f6f' }}>
          Type 'string' is not assignable to type 'number'.ts(2322) App.tsx(23,
          32): The expected type comes from property 'brand' which is declared
          here on type 'car: string; brand: number;'
        </p>

        {vegetables.map((veggie) => (
          <p key={veggie.name}>
            {veggie.name} ${veggie.price ? veggie.price : 'N/A'}
          </p>
        ))}
      </section>

      <section>
        <h2>Functions receiving objects as params</h2>
        <p>{JSON.stringify(createEmployee({ id: 1 }))}</p>
        <p>{JSON.stringify(createEmployee({ id: 2 }))}</p>
      </section>
    </div>
  );
}

export default App;
