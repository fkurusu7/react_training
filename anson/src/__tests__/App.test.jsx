// import { within } from "@testing-library/dom";
import { render, screen } from "@testing-library/react";
import { describe, expect, it } from "vitest";

import App from "../App";

describe("App", () => {
  /* it("should make a snapshot ", () => {
    const { container } = render(<App usersData={[]} />);
    expect(container).toMatchSnapshot();
  });

  describe("When there is only 1 user", () => {
    describe("Edit button is clicked", () => {
      it("should render save button", async () => {
        render(
          <App
            usersData={[
              {
                id: 1,
                name: "baru",
                email: "baru@barudesu.com",
                favoriteFood: ["sushi", "pizza"],
              },
            ]}
          />
        );

        const editButton = screen.getByRole("button", { name: "Edit" });
        await userEvent.click(editButton);

        const saveButton = screen.getByRole("button", { name: "Save" });
        expect(saveButton).exist;
      });

      it("should display user name & email input fields", async () => {
        render(
          <App
            usersData={[
              {
                id: 1,
                name: "baru",
                email: "baru@barudesu.com",
                favoriteFood: ["sushi", "pizza"],
              },
            ]}
          />
        );
        const editButton = screen.getByRole("button", { name: "Edit" });
        await userEvent.click(editButton);

        expect(screen.getByRole("textbox", { name: "name" })).exist;
        expect(screen.getByRole("textbox", { name: "email" })).exist;
      });
    });
  });

  describe("When there are 2 users", () => {
    it("shoudl have two users", () => {
      render(
        <App
          usersData={[
            {
              id: 1,
              name: "baru",
              email: "baru@barudesu.com",
              favoriteFood: ["sushi", "pizza"],
            },
            {
              id: 2,
              name: "kuru",
              email: "kuru@barudesu.com",
              favoriteFood: ["pizza"],
            },
          ]}
        />
      );

      expect(screen.getByText("baru")).exist;
      expect(screen.getByText("kuru")).exist;
    });

    it("should click edit button for 1st user and show save button", async () => {
      render(
        <App
          usersData={[
            {
              id: 1,
              name: "baru",
              email: "baru@barudesu.com",
              favoriteFood: ["sushi", "pizza"],
            },
            {
              id: 2,
              name: "kuru",
              email: "kuru@barudesu.com",
              favoriteFood: ["pizza"],
            },
          ]}
        />
      );

      const userDetails = screen.getByTestId("user-item-1");
      expect(within(userDetails).queryByText("kuru")).toBeNull;
      const editButton = within(userDetails).getByRole("button", {
        name: "Edit",
      });
      await userEvent.click(editButton);
      expect(within(userDetails).getByRole("button", { name: "Save" }));
    });

    it("should edit 2nd username and save", async () => {
      render(
        <App
          usersData={[
            {
              id: 1,
              name: "baru",
              email: "baru@barudesu.com",
              favoriteFood: ["sushi", "pizza"],
            },
            {
              id: 2,
              name: "kuru",
              email: "kuru@barudesu.com",
              favoriteFood: ["pizza"],
            },
          ]}
        />
      );

      const userDetails = screen.getByTestId("user-item-2");
      expect(within(userDetails).queryByText("baru")).toBeNull;

      // click button
      await userEvent.click(
        within(userDetails).getByRole("button", { name: "Edit" })
      );

      await userEvent.clear(
        within(userDetails).getByRole("textbox", { name: "name" }),
        "kuru"
      );
      await userEvent.type(
        within(userDetails).getByRole("textbox", { name: "name" }),
        "kurusito"
      );
      await userEvent.click(
        within(userDetails).getByRole("button", { name: "Save" })
      );

      expect(within(userDetails).queryByText("kuru")).toBeNull();
      expect(within(userDetails).getByText("kurusito")).exist;
    });
  }); */

  describe("rendering context data", () => {
    it("should match snapshot", () => {});

    it("should render correct email", async () => {
      const { container } = render(<App />);
      expect(await screen.findByText("Email: mina@mina.com"));
      expect(container).toMatchSnapshot();
    });
  });
});
