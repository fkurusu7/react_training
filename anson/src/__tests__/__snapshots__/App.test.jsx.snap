// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > rendering context data > should render correct email 1`] = `
<div>
  <div>
    <h1>
      JSON PlaceHolder Users
    </h1>
    <div>
      <div>
        <span>
          ID: 
        </span>
        <h2>
          Name: 
          mina
        </h2>
        <p>
          Username: 
          mineapolis
        </p>
        <p>
          Email: 
          mina@mina.com
        </p>
      </div>
      <p />
    </div>
    <div>
      Posts
    </div>
  </div>
</div>
`;
